File: lib/example/cplusplus/fun_abort_xam.cpp
// -----------------------------------------------------------------------------
//         cppad_py: A C++ Object Library and Python Interface to Cppad
//          Copyright (C) 2017-20 Bradley M. Bell (bradbell@seanet.com)
//              This program is distributed under the terms of the
//              GNU General Public License version 3.0 or later see
//                    https://www.gnu.org/licenses/gpl-3.0.txt
// -----------------------------------------------------------------------------
// abort_recording
// -----------------------------------------------------------------------------
// BEGIN SOURCE
# include <cstdio>
# include <cppad/py/cppad_py.hpp>

bool fun_abort_xam(void) {
	using cppad_py::a_double;
	using cppad_py::vec_double;
	using cppad_py::vec_a_double;
	using cppad_py::d_fun;
	//
	// initialize return variable
	bool ok = true;
	//------------------------------------------------------------------------
	int n_ind = 2;
	//
	// create ax
	vec_double x(n_ind);
	for(int i = 0; i < n_ind ; i++) {
		x[i] = i + 1.0;
	}
	vec_a_double ax = cppad_py::independent(x);
	//
	// preform some a_double operations
	a_double ax0 = ax[0];
	a_double ax1 = ax[1];
	a_double ay = ax0 + ax1;
	//
	// check that ay is a variable; its value depends on the value of ax
	ok = ok && ay.variable();
	//
	// abort this recording
	cppad_py::abort_recording();
	//
	// check that ay is now a parameter, no longer a variable.
	ok = ok && ay.parameter();
	//
	// since it is a parameter, we can retrieve its value
	double y = ay.value();
	//
	// its value should be x0 + x1
	ok = ok && y  == x[0] + x[1];
	//
	// an abort when not recording has no effect
	cppad_py::abort_recording();
	//
	return( ok );
}
// END SOURCE
//
/*
$begin fun_abort_xam.cpp$$
$spell
	cplusplus
	cppad
	py
	xam
	Jacobian
	Jacobians
$$
$section C++: Abort Recording a_double Operations: Example and Test$$
$srcthisfile|0|// BEGIN SOURCE|// END SOURCE|$$
$end
*/
//

File: example/python/core/fun_abort_xam.py
# SPDX-License-Identifier: GPL-3.0-or-later
# SPDX-FileCopyrightText: Bradley M. Bell <bradbell@seanet.com>
# SPDX-FileContributor: 2017-23 Bradley M. Bell
# ----------------------------------------------------------------------------
# abort_recording
# -----------------------------------------------------------------------------
# BEGIN SOURCE
def fun_abort_xam() :
   #
   import numpy
   import cppad_py
   #
   # initialize return variable
   ok = True
   # ---------------------------------------------------------------------
   n_ind = 2
   #
   # create ax
   x = numpy.empty(n_ind, dtype=float)
   for i in range( n_ind  ) :
      x[i] = i + 1.0
   #
   ax = cppad_py.independent(x)
   #
   # preform some a_double operations
   ax0 = ax[0]
   ax1 = ax[1]
   ay = ax0 + ax1
   #
   # check that ay is a variable; its value depends on the value of ax
   ok = ok and ay.variable()
   #
   # abort this recording
   cppad_py.abort_recording()
   #
   # check that ay is now a parameter, no longer a variable.
   ok = ok and ay.parameter()
   #
   # since it is a parameter, we can retrieve its value
   y = ay.value()
   #
   # its value should be x0 + x1
   ok = ok and y  == x[0] + x[1]
   #
   # an abort when not recording has no effect
   cppad_py.abort_recording()
   #
   return( ok )
#
# END SOURCE
#
#
# {xrst_begin fun_abort_xam.py}
# {xrst_comment_ch #}
#
# Python: Abort Recording a_double Operations: Example and Test
# #############################################################
# {xrst_literal
#  # BEGIN SOURCE
#  # END SOURCE
# }
# {xrst_end fun_abort_xam.py}
#
def test_fun_abort_xam() :
   assert fun_abort_xam()
